[CRM] Make Automations Workflow endpoints functional

kallehauge opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to projects/plugins/jetpack/to-test.md in a new commit as part of your PR. -->
<!-- a12s: If you have an expected version that you're aiming for the PR to add, please use the Milestone field to communicate it. If you leave it blank, that indicates there isn't a preference. -->

This PR will update the "Automation Workflows endpoints" that was previously added with dummy data, to being fully functional now. It aims to do so by:

* Complete the `[GET] /wp-json/jetpack-crm/v4/automation/workflows` endpoint
  * Params: `(bool) active`, `(string) category`, `(int) page`, `(int) per_page`, `(int) offset`.
* Complete the `[GET] /wp-json/jetpack-crm/v4/automation/workflows/{id}` endpoint
  * This only accepts the ID in the URL. 
* Complete the `[PUT] /wp-json/jetpack-crm/v4/automation/workflows/{id}` endpoint
  * Params: `(string) name`, `(string) description`, `(bool) active`, `(string) category`, `(string|int) initial_step`, `(object) steps`.
    * The "Steps" param is registered as an "object" because WP REST assumes "array" is a numeric array but we want to allow both strings and integers as the key. We also require a "slug" to be present as a simple verification that we're getting the correct format.

Fixes https://github.com/Automattic/zero-bs-crm/issues/3226

## Proposed changes:
<!--- Explain what functional changes your PR includes -->
* Rename controller from `REST_Automation_Controller` to `REST_Automation_Workflows_Controller` since, theoretically, if we add other endpoints than workflows, then it makes sense to separate them to improve readability and maintainability.
    * I've also renamed the `REST_Automation_Controller_Test` test class to `REST_Automation_Workflows_Controller_Test` to reflect the above change.
* Renamed methods inside the workflow controller to match default WordPress naming (`get_item`, `get_items`, etc).
* Added pagination to `Workflow_Repository::find_by`
* Changed the singular workflow paths to follow RESTful naming (`/jetpack-crm/v4/automation/workflow` => `/jetpack-crm/v4/automation/workflows/{id}`)
* Complete the workflow endpoints with dummy data:
    * Complete the `[GET] /wp-json/jetpack-crm/v4/automation/workflows` endpoint
    * Complete the `[GET] /wp-json/jetpack-crm/v4/automation/workflows/{id}` endpoint
    * Complete the `[PUT] /wp-json/jetpack-crm/v4/automation/workflows/{id}` endpoint
* Updated tests to use the workflow repository instead of relying on the dummy data that existed in the endpoint skeleton versions.

**Side-note:**

I've tried to keep a link between the old controller and test files, but I cannot for the life of me get it to work right now. So to save time while nearing the ETA, I've just chosen to accept it as is.

### Other information:

- [x] Have you written new tests for your changes, if applicable?
- [x] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->

* pbhBOz-3ss-p2
* pbhBOz-2ES-p2
* pbhBOz-3uX-p2

## Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->

No.

## Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

### Automated tests

All of the tests should obviously pass, but looking at the following automated tests should suffice:

* `REST_Authentication_Test`: Updated to include our new endpoints.
* `REST_Automation_Workflows_Controller_Test`: Updated to use real integration tests + added new tests to verify additional expected behaviour like 404 response(s).
* `Workflow_Repository_Test`: Updated to test limit/offset arguments when fetching workflows in bulk.

### Manual testing

**Manual testing prep work**

1. Make sure the following feature flags are enabled on your site:

```php
add_filter( 'jetpack_crm_feature_flag_api_v4', '__return_true' );
add_filter( 'jetpack_crm_feature_flag_automations', '__return_true' );
```

2. Modify `REST_Automation_Workflows_Controller::get_item_permissions_check` to return `true` at the top.

_We do this to make it easier to call the endpoints. This should allow you to do it from any client you want (e.g. Postman) without having to worry about WP Logged in cookies, nonce, etc._

3. Make sure you have a workflow in your database. You can use the below snippet to add one.

```php
$workflow = new \Automattic\Jetpack\CRM\Automation\Automation_Workflow(
	[
		'name'         => 'New contact: Set status to "Customer"',
		'description'  => 'This automation will change the status of a contact to "Customer" when they are created.',
		'category'     => 'contact',
		'active'       => true,
		'triggers'     => [
			\Automattic\Jetpack\CRM\Automation\Triggers\Contact_Created::get_slug(),
		],
		'initial_step' => 1,
		'steps'        => [
			1 => [
				'slug'       => \Automattic\Jetpack\CRM\Automation\Actions\Update_Contact_Status::get_slug(),
				'attributes' => [
					'new_status' => 'Customer',
				],
				'next_step'  => null,
			],
		],
	]
);

$repo = new \Automattic\Jetpack\CRM\Automation\Workflow\Workflow_Repository();
$workflow = $repo->persist(  $workflow );
```

**GET All Workflows**

* Method: `GET`
* Path: `/wp-json/jetpack-crm/v4/automation/workflows`
* Payload: N/A

**GET Workflow**

* Method: `GET`
* Path: `/wp-json/jetpack-crm/v4/automation/workflows/{id}`
* Payload: N/A

**Update Workflow**

* Method: `PUT` (or `POST`, both works)
* Path: `/wp-json/jetpack-crm/v4/automation/workflows/{id}`
* Payload: `{ "name": "My updated name" }`

kallehauge committed:
CRM: Add workflow create table query and migration script

Rebased onto trunk. Fixed conflicts.

kallehauge committed:
CRM: Add workflow properties so it matches the table columns in the database

kallehauge committed:
CRM: Add a repository for automation workflows

Resolving conflicts 2.

kallehauge committed:
CRM: Load active automation workflows

Resolving conflicts 3

kallehauge committed:
Added changelog

kallehauge committed:
CRM: Force workflow::active to be an integer when comparing

We have to do this since there's a change we get a numeric string.

cleacos committed:
Remove duplicated properties in Automation_Workflow.

Signed-off-by: cleacos <javacodia@gmail.com>

cleacos committed:
Set strict array and default empty in find_all method.

Signed-off-by: cleacos <javacodia@gmail.com>

cleacos committed:
Use boolean for active.

Signed-off-by: cleacos <javacodia@gmail.com>

cleacos committed:
Fix linter issue with the space

Signed-off-by: cleacos <javacodia@gmail.com>

cleacos committed:
Fix active value type, use bool instead of int

Signed-off-by: cleacos <javacodia@gmail.com>

kallehauge committed:
[CRM] Rename Workflow API methods and use repository

* Rename the methods to follow WordPress standards
* Use Workflow_Repository to fetch workflows
* Initial refactor of tests to support it

kallehauge committed:
Merge branch 'trunk' into crm/update/automation-workflow-endpoints

kallehauge committed:
Update workflows endpoint to be functional

kallehauge committed:
Merge branch 'trunk' into crm/update/automation-workflow-endpoints

kallehauge committed:
changelog

kallehauge committed:
Restore old endpoint file to create Git link

kallehauge committed:
Update controller

kallehauge committed:
Delete "new" controller with undercover name

kallehauge commented on projects/plugins/crm/tests/php/rest-api/v4/class-rest-authentication-test.php:
```diff
@@ -21,10 +21,22 @@ class REST_Authentication_Test extends REST_Base_Test_Case {
 				WP_REST_Server::READABLE,
 				'/jetpack-crm/v4/contacts/123',
 			),
+			'automation_workflows::get_items'   => array(
+				WP_REST_Server::READABLE,
+				'/jetpack-crm/v4/automation/workflows',
+			),
+			'automation_workflows::get_item'    => array(
+				WP_REST_Server::READABLE,
+				'/jetpack-crm/v4/automation/workflows/123',
+			),
+			'automation_workflows::update_item' => array(
+				WP_REST_Server::CREATABLE,
+				'/jetpack-crm/v4/automation/workflows/123',
+			),
```
This dataprovider is a collection of endpoints that require the logged in cookie to access them.
By adding our new endpoints here, we automatically get some auth tests.

cleacos commented on projects/plugins/crm/src/automation/workflow/class-workflow-repository.php:
```diff
@@ -117,6 +114,15 @@ public function find_by( array $criteria ): array {
 		}
 
+		// Build pagination condition.
+		if ( isset( $criteria['page'] ) || isset( $criteria['per_page'] ) || isset( $criteria['offset'] ) ) {
```
Better if we add the pagination/order functionality as params in the method:

`public function find_by( array $criteria, $order_by = 'id', int $limit = null, int $offset = 0 ): array {`

The reason is that we could have columns with those names: `page`, `per_page`, and `offset`.

I prefer the `limit` and `offset` terms to match with the SQL clauses.

By the way, `offset` and `page` would be the same.

cleacos replied:
But it's true that in this case, we are limiting the criteria allowed keys. But if we are going to generalize it, maybe is better to set it as params from now.

kallehauge replied:
We can definitely change the naming inside the repository itself.
The reason I went with that naming scheme is due to the REST API docs to follow best practice, but those can still be upheld in the endpoint controller instead: https://developer.wordpress.org/rest-api/using-the-rest-api/pagination/#pagination-parameters

I'll give it a look shortly

kallehauge replied:
>By the way, offset and page would be the same.

Agreed. There's a comment that also explains it for the code itself:

https://github.com/Automattic/jetpack/pull/33296/files#diff-df77d7f24c7769b6989fff4eb24d14d497b61c055a2971875cd46af9d5df6d7cR306-R309

kallehauge committed:
Refactor pagination logic to separate REST and underlying arguments

kallehauge committed:
Add REST pagination tests

kallehauge committed:
Update min/max docs link

kallehauge committed:
Add empty workflows result test

kallehauge committed:
GET Workflow: Add 404 test

kallehauge committed:
Update workflows docblocks

kallehauge committed:
Complete "Update workflow" endpoint

kallehauge committed:
Fix issue where nested "slug" property was not defined correctly

This meant that we never verified if it was present or not.

kallehauge committed:
Merge branch 'trunk' into crm/update/automation-workflow-endpoints

kallehauge committed:
Add "todo" with missing order by support.

We are adding the parameter there because we know it will be supported in the future, but it does not make sense to have it
after "limit" and "offset".

kallehauge committed:
[CRM] Add "Delete Workflow" endpoint (#33322)

* [CRM] Add "Delete Workflow" endpoint

* Add auth test

kallehauge committed:
Revert "[CRM] Add "Delete Workflow" endpoint (#33322)" (#33324)

This reverts commit 1986e0f2a51ee026a7649c72758b9b6c178a631f.

cleacos commented on projects/plugins/crm/src/automation/workflow/class-workflow-repository.php:
```diff
@@ -117,6 +120,22 @@ public function find_by( array $criteria ): array {
+			// It seems intuitive to provide "0" to mean "no limit", but technically it means that
+			// we do not want to return any results at all, so to help with the developer experience
+			// we convert "0" to a very high number instead.
+			if ( 0 === $limit ) {
+				$limit = 9999999999999;
+			}
```
For my business case, `9999999999999` is not enough :^)

And if we use `PHP_INT_MAX`?

tbradsha replied:
> :^)

@cleacos Imposter! But I agree, `PHP_INT_MAX` is better.

cleacos commented on projects/plugins/crm/tests/php/rest-api/v4/class-rest-automation-workflows-controller-test.php:
```diff
@@ -0,0 +1,307 @@
+	 *
+	 * @return void
+	 */
+	public function test_get_workflow_return_404_if_id_do_not_exist() {
```
👍

cleacos reviewed (dismissed):
The code looks good. Tests pass. The 3 manual tests work as expected, I can get all the workflows, get one workflow by the ID, and update the workflow.

**Bonus check:** I also tested to query workflows by filtering with other fields, and it works perfectly!!

- `/wp-json/jetpack-crm/v4/automation/workflows?active=0` => Not return anything, as it was expected (by my workflows in my DB)
- `/wp-json/jetpack-crm/v4/automation/workflows?active=1` => It returns the active workflows.
- `/wp-json/jetpack-crm/v4/automation/workflows?name=My updated name` => It returns the expected workflow.

LGTM!

kallehauge committed:
Merge branch 'trunk' into crm/update/automation-workflow-endpoints

kallehauge committed:
Use PHP_INT_MAX instead of an arbitrary number
